src/analyses/useAfterFree.ml

```diff
@@ -0,0 +1,153 @@
+      warn_exp_might_contain_freed ~is_double_free transfer_fn_name e3 ctx
+    (* Lval cases (need [warn_lval_might_contain_freed] for them) *)
+    | Lval lval
+    | StartOf lval
+    | AddrOf lval -> warn_lval_might_contain_freed ~is_double_free transfer_fn_name lval ctx
+
+
+  (* TRANSFER FUNCTIONS *)
+
+  let assign ctx (lval:lval) (rval:exp) : D.t =
+    warn_lval_might_contain_freed "assign" lval ctx;
+    warn_exp_might_contain_freed "assign" rval ctx;
+    ctx.local
+
+  let branch ctx (exp:exp) (tv:bool) : D.t =
+    warn_exp_might_contain_freed "branch" exp ctx;
+    ctx.local
+
+  let body ctx (f:fundec) : D.t =
+    ctx.local
+
+  let return ctx (exp:exp option) (f:fundec) : D.t =
+    Option.iter (fun x -> warn_exp_might_contain_freed "return" x ctx) exp;
+    ctx.local
+
+  let enter ctx (lval:lval option) (f:fundec) (args:exp list) : (D.t * D.t) list =
+    let caller_state = ctx.local in
+    Option.iter (fun x -> warn_lval_might_contain_freed "enter" x ctx) lval;
+    List.iter (fun arg -> warn_exp_might_contain_freed "enter" arg ctx) args;
+    if D.is_empty caller_state then
+      [caller_state, caller_state]
+    else (
+      let reachable_from_args = List.fold_left (fun acc arg -> Queries.LS.join acc (ctx.ask (ReachableFrom arg))) (Queries.LS.empty ()) args in
+      if Queries.LS.is_top reachable_from_args || D.is_top caller_state then
+        [caller_state, caller_state]
+      else
+        let reachable_vars =
+          Queries.LS.elements reachable_from_args
+          |> List.map fst
+          |> List.filter (fun var -> ctx.ask (Queries.IsHeapVar var))
+          |> D.of_list
+        in
+        let callee_state = D.inter caller_state reachable_vars in
+        [caller_state, callee_state]
+    )
+
+  let combine_env ctx (lval:lval option) fexp (f:fundec) (args:exp list) fc (callee_local:D.t) (f_ask:Queries.ask) : D.t =
+    let caller_state = ctx.local in
+    D.join caller_state callee_local
+
+  let combine_assign ctx (lval:lval option) fexp (f:fundec) (args:exp list) fc (callee_local:D.t) (f_ask: Queries.ask): D.t =
+    ctx.local
+
+  let special ctx (lval:lval option) (f:varinfo) (arglist:exp list) : D.t =
+    let state = ctx.local in
+    Option.iter (fun x -> warn_lval_might_contain_freed ("special: " ^ f.vname) x ctx) lval;
+    List.iter (fun arg -> warn_exp_might_contain_freed ~is_double_free:(f.vname = "free") ("special: " ^ f.vname) arg ctx) arglist;
+    let desc = LibraryFunctions.find f in
+    match desc.special arglist with
+    | Free ptr ->
+      begin match ctx.ask (Queries.MayPointTo ptr) with
+        | a when not (Queries.LS.is_top a) && not (Queries.LS.mem (dummyFunDec.svar, `NoOffset) a) ->
+          begin try
+              let (v, _) = Queries.LS.choose a in
+              if ctx.ask (Queries.IsHeapVar v) then
+                D.add v state
+              else state
+            with Not_found -> state
+          end
```

Comment:
What if there are several things that the pointer could point to? We would have to iterate over all of them and add them to the set of may-freed things.

Reply:
Thanks for noting this! Should be fixed in https://github.com/goblint/analyzer/pull/1050/commits/47eff4483145b24366c8b80089ffd1e8663cb01f

(after https://github.com/goblint/analyzer/pull/1050/commits/47eff4483145b24366c8b80089ffd1e8663cb01f, Goblint should now also warn for each possible pointed-to heap var if it's used after being freed; earlier a warning was only issued for a single heap var, selected via `Queries.LS.choose`)